File: src/utils/__tests__/fqbn.test.js
/**
 * fqbn.test.js
 * Tests for the FQBN parser utilities.
 */

import { describe, it, expect } from 'vitest';
import {
  parseFQBN, serializeFQBN, isValidFQBN, normalizeFQBN,
  platformID, friendlyName, mcuFromFQBN,
} from '../fqbn';

// ── parseFQBN ─────────────────────────────────────────────────

describe('parseFQBN', () => {
  it('parses basic 3-segment FQBN', () => {
    const result = parseFQBN('arduino:avr:uno');
    expect(result).toMatchObject({ vendor: 'arduino', architecture: 'avr', board: 'uno', options: {} });
  });

  it('parses FQBN with menu options', () => {
    const result = parseFQBN('arduino:avr:mega:cpu=atmega2560');
    expect(result.options).toEqual({ cpu: 'atmega2560' });
  });

  it('parses FQBN with multiple menu options', () => {
    const result = parseFQBN('esp32:esp32:esp32:freq=240,psram=enabled');
    expect(result.options).toEqual({ freq: '240', psram: 'enabled' });
  });

  it('returns null for empty string', () => {
    expect(parseFQBN('')).toBeNull();
    expect(parseFQBN(null)).toBeNull();
  });

  it('returns null for incomplete FQBN (< 3 segments)', () => {
    expect(parseFQBN('arduino:avr')).toBeNull();
    expect(parseFQBN('arduino')).toBeNull();
  });

  it('handles STMicro long vendor name', () => {
    const result = parseFQBN('STMicroelectronics:stm32:Nucleo_64');
    expect(result.vendor).toBe('STMicroelectronics');
    expect(result.board).toBe('Nucleo_64');
  });
});

// ── serializeFQBN ─────────────────────────────────────────────

describe('serializeFQBN', () => {
  it('serializes basic FQBN', () => {
    expect(serializeFQBN({ vendor: 'arduino', architecture: 'avr', board: 'uno', options: {} }))
      .toBe('arduino:avr:uno');
  });

  it('serializes with options — sorted alphabetically', () => {
    const result = serializeFQBN({
      vendor: 'esp32', architecture: 'esp32', board: 'esp32',
      options: { psram: 'enabled', freq: '240' },
    });
    expect(result).toBe('esp32:esp32:esp32:freq=240,psram=enabled');
  });
});

// ── normalizeFQBN ─────────────────────────────────────────────

describe('normalizeFQBN', () => {
  it('normalizes FQBN with unsorted options', () => {
    const result = normalizeFQBN('esp32:esp32:esp32:psram=enabled,freq=240');
    expect(result).toBe('esp32:esp32:esp32:freq=240,psram=enabled');
  });

  it('passes through valid FQBN unchanged (no options)', () => {
    expect(normalizeFQBN('arduino:avr:uno')).toBe('arduino:avr:uno');
  });

  it('returns original string for invalid FQBN', () => {
    expect(normalizeFQBN('not-an-fqbn')).toBe('not-an-fqbn');
  });
});

// ── isValidFQBN ───────────────────────────────────────────────

describe('isValidFQBN', () => {
  it('validates correct FQBNs',    () => { expect(isValidFQBN('arduino:avr:uno')).toBe(true); expect(isValidFQBN('esp32:esp32:esp32s3')).toBe(true); });
  it('rejects incomplete FQBNs',   () => { expect(isValidFQBN('arduino:avr')).toBe(false); });
  it('rejects empty string',       () => { expect(isValidFQBN('')).toBe(false); expect(isValidFQBN(null)).toBe(false); });
});

// ── platformID ────────────────────────────────────────────────

describe('platformID', () => {
  it('extracts vendor:architecture', () => {
    expect(platformID('arduino:avr:uno')).toBe('arduino:avr');
    expect(platformID('esp32:esp32:esp32:freq=240')).toBe('esp32:esp32');
  });
  it('returns null for invalid', () => { expect(platformID('bad')).toBeNull(); });
});

// ── friendlyName ──────────────────────────────────────────────

describe('friendlyName', () => {
  it('returns known board names',    () => { expect(friendlyName('arduino:avr:uno')).toBe('Arduino Uno'); expect(friendlyName('esp32:esp32:esp32')).toBe('ESP32 Dev Module'); });
  it('returns board segment for unknown', () => { expect(friendlyName('custom:hw:myboard')).toBe('myboard'); });
  it('appends (custom) for boards with options', () => { expect(friendlyName('arduino:avr:mega:cpu=atmega2560')).toContain('custom'); });
  it('handles null/empty gracefully', () => { expect(friendlyName('')).toBe(''); expect(friendlyName(null)).toBe(''); });
});

// ── mcuFromFQBN ───────────────────────────────────────────────

describe('mcuFromFQBN', () => {
  it('maps arduino:avr → avr',        () => { expect(mcuFromFQBN('arduino:avr:uno')).toBe('avr');    });
  it('maps esp32:esp32 → esp32',      () => { expect(mcuFromFQBN('esp32:esp32:esp32')).toBe('esp32'); });
  it('maps esp8266:esp8266 → esp8266',() => { expect(mcuFromFQBN('esp8266:esp8266:nodemcuv2')).toBe('esp8266'); });
  it('maps stm32 → stm32',            () => { expect(mcuFromFQBN('STMicroelectronics:stm32:Nucleo_64')).toBe('stm32'); });
  it('returns generic for unknown',   () => { expect(mcuFromFQBN('custom:hw:board')).toBe('generic'); });
});

File: companion-cli/internal/fqbn/fqbn_test.go
package fqbn_test

import (
	"testing"

	"github.com/companion-ide/companion-cli/internal/fqbn"
)

// ── Parse ──────────────────────────────────────────────────────

func TestParse_Basic(t *testing.T) {
	f, err := fqbn.Parse("arduino:avr:uno")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if f.Vendor != "arduino" || f.Architecture != "avr" || f.Board != "uno" {
		t.Errorf("got %+v", f)
	}
	if len(f.Options) != 0 {
		t.Errorf("expected no options, got %v", f.Options)
	}
}

func TestParse_WithOptions(t *testing.T) {
	f, err := fqbn.Parse("arduino:avr:mega:cpu=atmega2560")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if f.Options["cpu"] != "atmega2560" {
		t.Errorf("expected cpu=atmega2560, got %v", f.Options)
	}
}

func TestParse_MultipleOptions(t *testing.T) {
	f, err := fqbn.Parse("esp32:esp32:esp32:freq=240,psram=enabled")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if f.Options["freq"] != "240" || f.Options["psram"] != "enabled" {
		t.Errorf("got options: %v", f.Options)
	}
}

func TestParse_Empty(t *testing.T) {
	if _, err := fqbn.Parse(""); err == nil {
		t.Error("expected error for empty string")
	}
}

func TestParse_TwoSegments(t *testing.T) {
	if _, err := fqbn.Parse("arduino:avr"); err == nil {
		t.Error("expected error for 2-segment FQBN")
	}
}

func TestParse_OneSegment(t *testing.T) {
	if _, err := fqbn.Parse("arduino"); err == nil {
		t.Error("expected error for 1-segment FQBN")
	}
}

func TestParse_STMicro(t *testing.T) {
	f, err := fqbn.Parse("STMicroelectronics:stm32:Nucleo_64")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if f.Vendor != "STMicroelectronics" || f.Board != "Nucleo_64" {
		t.Errorf("got %+v", f)
	}
}

// ── String / Serialize ─────────────────────────────────────────

func TestString_NoOptions(t *testing.T) {
	f := fqbn.MustParse("arduino:avr:uno")
	if got := f.String(); got != "arduino:avr:uno" {
		t.Errorf("expected arduino:avr:uno, got %q", got)
	}
}

func TestString_OptionsAreSorted(t *testing.T) {
	// Options should be alphabetically sorted regardless of parse order
	f, _ := fqbn.Parse("esp32:esp32:esp32:psram=enabled,freq=240")
	got := f.String()
	want := "esp32:esp32:esp32:freq=240,psram=enabled"
	if got != want {
		t.Errorf("expected %q, got %q", want, got)
	}
}

// ── PlatformID ─────────────────────────────────────────────────

func TestPlatformID(t *testing.T) {
	f := fqbn.MustParse("arduino:avr:uno")
	if got := f.PlatformID(); got != "arduino:avr" {
		t.Errorf("expected arduino:avr, got %q", got)
	}
}

// ── Equal / EqualCore ─────────────────────────────────────────

func TestEqual(t *testing.T) {
	a := fqbn.MustParse("arduino:avr:mega:cpu=atmega2560")
	b := fqbn.MustParse("arduino:avr:mega:cpu=atmega2560")
	c := fqbn.MustParse("arduino:avr:mega:cpu=atmega1280")
	if !a.Equal(b) {
		t.Error("equal FQBNs should be equal")
	}
	if a.Equal(c) {
		t.Error("different option FQBNs should not be equal")
	}
}

func TestEqualCore(t *testing.T) {
	a := fqbn.MustParse("arduino:avr:mega:cpu=atmega2560")
	b := fqbn.MustParse("arduino:avr:mega:cpu=atmega1280")
	if !a.EqualCore(b) {
		t.Error("same board different options should be EqualCore")
	}
}

// ── IsValid ────────────────────────────────────────────────────

func TestIsValid(t *testing.T) {
	cases := []struct {
		raw  string
		want bool
	}{
		{"arduino:avr:uno", true},
		{"esp32:esp32:esp32:freq=240", true},
		{"STMicroelectronics:stm32:Nucleo_64", true},
		{"arduino:avr", false},
		{"", false},
		{"bad", false},
	}
	for _, tc := range cases {
		got := fqbn.IsValid(tc.raw)
		if got != tc.want {
			t.Errorf("IsValid(%q) = %v, want %v", tc.raw, got, tc.want)
		}
	}
}

// ── Normalize ──────────────────────────────────────────────────

func TestNormalize_SortsOptions(t *testing.T) {
	in := "esp32:esp32:esp32:psram=enabled,freq=240"
	want := "esp32:esp32:esp32:freq=240,psram=enabled"
	if got := fqbn.Normalize(in); got != want {
		t.Errorf("expected %q, got %q", want, got)
	}
}

func TestNormalize_PassthroughInvalid(t *testing.T) {
	in := "not-an-fqbn"
	if got := fqbn.Normalize(in); got != in {
		t.Errorf("expected passthrough of invalid FQBN, got %q", got)
	}
}

// ── FriendlyName ──────────────────────────────────────────────

func TestFriendlyName(t *testing.T) {
	cases := []struct {
		raw  string
		want string
	}{
		{"arduino:avr:uno", "Arduino Uno"},
		{"esp32:esp32:esp32", "ESP32 Dev Module"},
		{"custom:hw:myboard", "myboard"},
	}
	for _, tc := range cases {
		if got := fqbn.FriendlyName(tc.raw); got != tc.want {
			t.Errorf("FriendlyName(%q) = %q, want %q", tc.raw, got, tc.want)
		}
	}
}

// ── MCUFromFQBN ───────────────────────────────────────────────

func TestMCUFromFQBN(t *testing.T) {
	cases := []struct{ raw, want string }{
		{"arduino:avr:uno", "avr"},
		{"esp32:esp32:esp32", "esp32"},
		{"esp8266:esp8266:nodemcuv2", "esp8266"},
		{"STMicroelectronics:stm32:Nucleo_64", "stm32"},
		{"custom:hw:board", "generic"},
	}
	for _, tc := range cases {
		if got := fqbn.MCUFromFQBN(tc.raw); got != tc.want {
			t.Errorf("MCUFromFQBN(%q) = %q, want %q", tc.raw, got, tc.want)
		}
	}
}
